dpcpp/matrix/hybrid_kernels.dp.cpp

```diff
@@ -57,6 +65,172 @@ namespace dpcpp {
+                 const IndexType *__restrict__ ell_col,
+                 const ValueType *__restrict__ coo_val,
+                 const IndexType *__restrict__ coo_col,
+                 const IndexType *__restrict__ coo_offset,
+                 IndexType *__restrict__ result_row_ptrs,
+                 IndexType *__restrict__ result_col_idxs,
+                 ValueType *__restrict__ result_values,
+                 sycl::nd_item<3> item_ct1)
+{
+    const auto tidx = thread::get_thread_id_flat(item_ct1);
+
+    if (tidx < num_rows) {
+        auto write_to = result_row_ptrs[tidx];
+        for (size_type i = 0; i < max_nnz_per_row; i++) {
+            const auto source_idx = tidx + stride * i;
+            if (ell_val[source_idx] != zero<ValueType>()) {
+                result_values[write_to] = ell_val[source_idx];
+                result_col_idxs[write_to] = ell_col[source_idx];
+                write_to++;
+            }
+        }
+        for (auto i = coo_offset[tidx]; i < coo_offset[tidx + 1]; i++) {
+            if (coo_val[i] != zero<ValueType>()) {
+                result_values[write_to] = coo_val[i];
+                result_col_idxs[write_to] = coo_col[i];
+                write_to++;
+            }
+        }
+    }
+}
+
+template <typename ValueType, typename IndexType>
+void fill_in_csr(dim3 grid, dim3 block, size_type dynamic_shared_memory,
+                 sycl::queue *queue, size_type num_rows,
+                 size_type max_nnz_per_row, size_type stride,
+                 const ValueType *ell_val, const IndexType *ell_col,
+                 const ValueType *coo_val, const IndexType *coo_col,
+                 const IndexType *coo_offset, IndexType *result_row_ptrs,
+                 IndexType *result_col_idxs, ValueType *result_values)
+{
+    queue->submit([&](sycl::handler &cgh) {
+        cgh.parallel_for(
+            sycl_nd_range(grid, block), [=](sycl::nd_item<3> item_ct1) {
+                fill_in_csr(num_rows, max_nnz_per_row, stride, ell_val, ell_col,
+                            coo_val, coo_col, coo_offset, result_row_ptrs,
+                            result_col_idxs, result_values, item_ct1);
+            });
+    });
+}
+
+
+template <typename ValueType1, typename ValueType2>
+void add(size_type num, ValueType1 *__restrict__ val1,
+         const ValueType2 *__restrict__ val2, sycl::nd_item<3> item_ct1)
+{
+    const auto tidx = thread::get_thread_id_flat(item_ct1);
+    if (tidx < num) {
+        val1[tidx] += val2[tidx];
+    }
+}
+
+template <typename ValueType1, typename ValueType2>
+void add(dim3 grid, dim3 block, size_type dynamic_shared_memory,
+         sycl::queue *queue, size_type num, ValueType1 *val1,
+         const ValueType2 *val2)
+{
+    queue->submit([&](sycl::handler &cgh) {
+        cgh.parallel_for(
+            sycl_nd_range(grid, block),
+            [=](sycl::nd_item<3> item_ct1) { add(num, val1, val2, item_ct1); });
```

Comment:
Nit: For small functions, you could probably just put the function body in the lambda here.